staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go

```diff
@@ -87,7 +87,7 @@ func (a customResourceStrategy) PrepareForCreate(ctx context.Context, obj runtim
 
 // PrepareForUpdate clears fields that are not allowed to be set by end users on update.
 func (a customResourceStrategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Object) {
-	if !utilfeature.DefaultFeatureGate.Enabled(apiextensionsfeatures.CustomResourceSubresources) || a.status == nil {
+	if !utilfeature.DefaultFeatureGate.Enabled(apiextensionsfeatures.CustomResourceSubresources) {
```

Comment:
> A CR can have the .status field, while opted-out of the /status endpoint. User would expect the metadata.generation to auto-increment.

My understanding is that `.metadata.generation` should be incremented only when the user specifically says that they want to follow spec/status conventions - which means they need to explicitly set the `.subresources.status` field in the CRD and opt-in to `/status`. Enabling the feature gate does not guarantee a status subresource because it could also mean that they only want to enable `/scale`.

Related: https://github.com/kubernetes/kubernetes/issues/58778#issuecomment-360315841

Reply:
@caesarxuchao Do you have a use case in mind for wanting `generation` to increment even if the CRD author did not explicitly opt into the full spec/status convention (by enabling /status subresource)?

We intentionally chose not to infer that the CRD author wants to use the spec/status convention just because they have fields named `spec` and/or `status`. That would make it impossible to use those field names without participating in the convention, which would be a backwards-incompatible change.

Or are you just saying it's a confusing UX that you have to turn on a subresource to opt into the spec/status convention? If that's the case, perhaps there could be a separate field to enable `generation` for `spec` without enabling the `/status` subresource, with validation that you can't have the subresource without having `generation` too. I'm not aware of a strong use case for wanting those things separately, though.

Reply:
Maybe users just want to see how many times the `spec` has been changed.

The explanation in the second paragraph makes sense.

It's a confusing UX that I have to turn on a subresource to make apiserver manages my `generation`. Adding a specific field to turn on generation management is better than the status quo. But why is it bad to always let the apiserver manage the generation field as long as the CR has the `spec` field?

Reply:
> I almost forgot I had this early involvement in the 58778 :)

:)

>  If that's the case, perhaps there could be a separate field to enable generation for spec without enabling the /status subresource, with validation that you can't have the subresource without having generation too.

IMHO adding a new field is too much complexity for this.

> But why is it bad to always let the apiserver manage the generation field as long as the CR has the spec field?

Thinking about this more, this makes sense but I still have one concern.

The current behaviour is to increment the generation if the user opts in to `/status` i.e. they opt-in to spec _and status_ conventions. We did this because the general expectation is that `metadata.generation` will be useful in a controller with a corresponding `status.observedGeneration` (which implies that the user should also opt-in to status). This is also what https://github.com/kubernetes/kubernetes/issues/58778#issuecomment-360315841 said.

But `metadata.generation` is technically a "per object sequence number" (taking the terminology from https://github.com/kubernetes/kubernetes/issues/7328). The sequence number of an object should _not_ depend on the presence of `status.observedGeneration`. It should increment with every spec change, irrespective of whether it will prove useful.

This means that we just require the user to opt-in to spec conventions. Opting in to status conventions should not be a requirement. `.spec` holds special meaning for both `/status` and `/scale`. So, if one of these is enabled, I think we should increment generation.

My only concern is that just enabling the feature without opting in to `/status` or `/scale` doesn't guarantee that I want to follow spec`conventions. Right now, enabling the feature gate does not hold much meaning unless the user specifically opts-in to either one of the subresources.

> That would make it impossible to use those field names without participating in the convention, which would be a backwards-incompatible change.

With this change, we redefine the feature gate to also mean that `.spec` will participate in the spec conventions. I'm concerned that it will be a backwards-incompatible change. 

(But this doesn't really help your use case :/)